protocol/daemons/server/types/update_monitor.go

```diff
@@ -0,0 +1,150 @@
+package types
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+type updateMetadata struct {
+	timer           *time.Timer
+	updateFrequency time.Duration
+}
+
+// UpdateMonitor monitors the update frequency of daemon services. If a daemon service does not respond within
+// the maximum acceptable update delay set when the daemon is registered, the monitor will panic and halt the
+// protocol. This was judged to be the best solution for network performance because it prevents any validator from
+// participating in the network at all if a daemon service is not responding.
+type UpdateMonitor struct {
+	// serviceToUpdateMetadata maps daemon service names to their update metadata.
+	serviceToUpdateMetadata map[string]updateMetadata
+	// stopped indicates whether the monitor has been stopped. Additional daemon services cannot be registered
+	// after the monitor has been stopped.
+	stopped bool
+	// disabled indicates whether the monitor has been disabled. This is used to disable the monitor in testApp
+	// tests, where app.New is not executed.
+	disabled bool
+	// lock is used to synchronize access to the monitor.
+	lock sync.Mutex
+}
+
+// NewUpdateFrequencyMonitor creates a new update frequency monitor.
+func NewUpdateFrequencyMonitor() *UpdateMonitor {
+	return &UpdateMonitor{
+		serviceToUpdateMetadata: make(map[string]updateMetadata),
+	}
+}
+
+func (ufm *UpdateMonitor) DisableForTesting() {
+	ufm.lock.Lock()
+	defer ufm.lock.Unlock()
+
+	ufm.disabled = true
+}
+
+// RegisterDaemonServiceWithCallback registers a new daemon service with the update frequency monitor. If the daemon
+// service fails to respond within the maximum acceptable update delay, the monitor will execute the callback function.
+// This method is synchronized. The method returns an error if the daemon service was already registered or the
+// monitor has already been stopped.
+func (ufm *UpdateMonitor) RegisterDaemonServiceWithCallback(
+	service string,
+	maximumAcceptableUpdateDelay time.Duration,
+	callback func(),
+) error {
+	ufm.lock.Lock()
+	defer ufm.lock.Unlock()
+
+	if maximumAcceptableUpdateDelay <= 0 {
+		return fmt.Errorf(
+			"registration failure for service %v: maximum acceptable update delay must be positive",
+			service,
+		)
+	}
+
+	// Don't register daemon services if the monitor has been disabled.
+	if ufm.disabled {
+		return nil
+	}
+
+	// Don't register additional daemon services if the monitor has already been stopped.
+	// This could be a concern for short-running integration test cases, where the network
+	// stops before all daemon services have been registered.
+	if ufm.stopped {
+		return fmt.Errorf("registration failure for service %v: monitor has been stopped", service)
+	}
+
+	_, ok := ufm.serviceToUpdateMetadata[service]
+	if ok {
```

Comment:
Suggestion, up to you:
```suggestion
	if _, ok := ufm.serviceToUpdateMetadata[service]; ok {
```